Engine API: Make SYNCING response and according actions fork specific (#152)
ethereum/execution-apis: src/engine/specification.md

    * If validation succeeds, return `{status: VALID, latestValidHash: payload.blockHash}`
    * If validation fails, return `{status: INVALID, latestValidHash: validHash}` where `validHash` is the block hash of the most recent *valid* ancestor of the invalid payload. That is, the valid ancestor of the payload with the highest `blockNumber`.

1. Client software **MUST** discard the payload if it's deemed invalid.

1. Client software **MUST** return `{status: SYNCING, latestValidHash: null}` if requisite data for payload validation is missing. In the event that requisite data to validate the payload is missing (e.g. does not have payload identified by `parentHash`), the client software **SHOULD** initiate the sync process.

> 1. Client software **MUST** validate the payload and respond accordingly despite of the sync process being in progress if there is enough data for the payload validation.

Reviewer: Might consider a clarifying example

"(e.g. if branch A is `SYNCING` but a block is inserted the head of branch B which is VALID, this block must be validated)"

1. The values `(forkchoiceState.headBlockHash, forkchoiceState.finalizedBlockHash)` of this method call map on the `POS_FORKCHOICE_UPDATED` event of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification) and **MUST** be processed according to the specification defined in the EIP.

1. All updates to the forkchoice state resulting from this call **MUST** be made atomically.

> 1. Client software **MUST** return `{status: SUCCESS, payloadId: null}` if `payloadAttributes` is `null`.

Reviewer: We should clarify that this isn't returned *no matter what* if `payloadAttribues` is `null` but only if null and successful

```suggestion
1. Client software **MUST** return `{status: SUCCESS, payloadId: null}` if `payloadAttributes` is `null` and fork choice update is successful.
```

Reviewer: Actually, maybe we should reorder this statement with the next one.

* `{SYNCING, null}` if not enough info to process
* `{SUCCESS, null}` if no payload and success
* `{SUCCESS, id}` if payload and success

Reviewer: That way it acts a blocking condition on the next two success conditions

> 1. Client software **MUST** return `{status: SYNCING, payloadId: null}` if the payload identified by either the `forkchoiceState.headBlockHash` or the `forkchoiceState.finalizedBlockHash` is unknown. In the event that either the `forkchoiceState.headBlockHash` or the `forkchoiceState.finalizedBlockHash` is unknown, the client software **SHOULD** initiate the sync process.

Reviewer: Should the **SHOULD** here be a **MUST**?

It's weird to respond with `SYNCING` and not actually kick off some sync process

Author: Data supplied by the `forkchoiceUpdated` isn't always enough to initiate the sync process. For instance, there is no state root to pull the state for, so, the **MUST** can't always be satisfied.

Reviewer: I see

1. Client software **MUST** return `{status: SUCCESS, payloadId: buildProcessId}` if `payloadAttributes` is not `null`, and **MUST** begin a payload build process building on top of `forkchoiceState.headBlockHash` and identified via `buildProcessId` value. The build process is specified in the [Payload build process](#payload-build-process) section.

> 1. Client software **MUST** update the fork choice state and initiate a payload build process, and respond accordingly despite of the sync process being in progress if payloads identified by the `forkchoiceState.headBlockHash` and the `forkchoiceState.finalizedBlockHash` are known.

Reviewer: ```suggestion
1. Client software **MUST** update the fork choice state and initiate a payload build process, and respond accordingly even if the sync process is in progress on some branch if payloads identified by the `forkchoiceState.headBlockHash` and the `forkchoiceState.finalizedBlockHash` are known.
```
